dfget/core/downloader/p2p_downloader/p2p_downloader.go

```diff
@@ -150,18 +154,40 @@ func (p2p *P2PDownloader) init() {
 
 // Run starts to download the file.
 func (p2p *P2PDownloader) Run(ctx context.Context) error {
+	if p2p.streamMode {
+		return fmt.Errorf("streamMode enabled, should be disable")
+	}
+	pieceProcessor := NewClientWriter(p2p.clientFilePath, p2p.serviceFilePath, p2p.clientQueue, p2p.API, p2p.cfg)
+	return p2p.run(ctx, pieceProcessor)
+}
+
+func (p2p *P2PDownloader) RunStream(ctx context.Context) (io.Reader, error) {
+	if !p2p.streamMode {
+		return nil, fmt.Errorf("streamMode enabled, should be disable")
+	}
+	pieceProcessor := NewClientStreamWriter(p2p.clientQueue, p2p.API, p2p.cfg)
+	go func() {
+		err := p2p.run(ctx, pieceProcessor)
+		if err != nil {
+			logrus.Warnf("P2PDownloader run error: %s", err)
+		}
+	}()
+	return pieceProcessor, nil
+}
+
+func (p2p *P2PDownloader) run(ctx context.Context, pieceProcessor PieceWriter) error {
 	var (
 		lastItem *Piece
 		goNext   bool
+		err      error
 	)
 
-	// start ClientWriter
-	clientWriter, err := NewClientWriter(p2p.clientFilePath, p2p.serviceFilePath, p2p.clientQueue, p2p.API, p2p.cfg)
-	if err != nil {
+	// start PieceWriter
+	if err = pieceProcessor.PreRun(ctx); err != nil {
```

Comment:
IMOP, use `err := pieceProcessor.PreRun(ctx)` is better.

Reply:
Done